_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,254 @@
+---
+title: "Digging into struct initialization performance"
+author: ysbaddaden
+summary: "How a quick test of BLAKE3 vs SHA256 in Crystal unveiled a performance issue that stems from the Ruby syntax."
+categories: technical
+tags: [language,performance]
+---
+
+I toyed with BLAKE3 some months ago. I wanted to verify how much faster was BLAKE3 versus SHA256 in Crystal, and see if it would bring some free improvement. Luckily there is [a shard](https://github.com/didactic-drunk/blake3.cr) wrapping the [official library](https://github.com/BLAKE3-team/BLAKE3) that is highly optimized with custom assembly for different CPU features (SSE, AVX2, NEON …).
+
+To my surprise, performance was pretty bad, and this post is the written legacy of the quest to understand why. Be prepared to dig into technical details and how language design can impact performance.
+
+## Let’s benchmark
+
+One of the points for BLAKE3 is that it’s magnitudes faster than alternative hash digests; the authors claim almost 14 times faster than SHA256 for example. I wrote a quick benchmark, comparing `Digest::Blake3` to `Digest::SHA256` (backed by OpenSSL) that is usually my default choice for my use cases, for example hash a session id made of 32 bytes.
+
+```crystal
+require "benchmark"
+require "blake3"
+require "digest/sha256"
+
+class Digest::Blake3 < ::Digest
+  extend ::Digest::ClassMethods
+end
+
+Benchmark.ips do |x|
+  bytes = Random::Secure.random_bytes(32)
+  x.report("SHA256") { Digest::SHA256.hexdigest(bytes) }
+  x.report("Blake3") { Digest::Blake3.hexdigest(bytes) }
+end
+```
+
+```
+SHA256   1.33M (753.33ns) (± 0.91%)    224B/op        fastest
+Blake3   1.21M (827.21ns) (± 0.79%)  2.13kB/op   1.10× slower
+```
+
+And the winner is… SHA256! What?
+
+## Let’s investigate
+
+The benchmark shows that `Digest::Blake3` allocates 2.13KB of memory in the HEAP for each iteration. Looking into the BLAKE3 algorithm, this is by design: the algorithm needs almost 2KB of state to compute the hash digest. That’s a lot of memory, and such a benchmark allocates memory just to throw it away immediately. Repeated HEAP allocations slow things down, as it puts pressure on the GC (it needs to regularly mark/sweep the memory which is a slow and blocking operation).
+
+> [!TIP]
```

Comment:
Callouts are described [in the guide](https://crystal-lang.org/style-guide/#typography). One option is to do the following:
```suggestion
> **NOTE:** Tip
```

Or add it to the list of callouts (see _config.yml)